### PR TITLE
Buffer AAD in wolfSSL_EVP_CipherUpdate_GCM  so that whole value is hashed

### PR DESCRIPTION
**Pull request by julek-wolfssl**

### CODE REVIEW

File: wolfcrypt/src/evp.c [hunk 1]
```diff
@@ -4123,6 +4161,11 @@ int wolfSSL_EVP_MD_type(const WOLFSSL_EVP_MD *md)
                 ctx->gcmDecryptBuffer = NULL;
             }
             ctx->gcmDecryptBufferLen = 0;
+            if (ctx->gcmAuthIn) {
+                XFREE(ctx->gcmAuthIn, NULL, DYNAMIC_TYPE_OPENSSL);
+                ctx->gcmAuthIn = NULL;
+            }
+            ctx->gcmAuthInSz = 0;
```
Comment (SparkiDev) [inline wolfcrypt/src/evp.c:4168]: Does reuse of the cipher reset the gcmAuthInSz?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by dgarske** [comment]
@julek-wolfssl : You may also need to update this area of code: https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/src/evp.c#L4093

See ZD 10931. @embhorn

File: wolfcrypt/src/evp.c (hunk 1 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/evp.c:4168]: If you do not call EVP_CIPHER_CTX_free then the same AAD is used for the next Init-Update-Final cycle.

File: tests/api.c [hunk 2]
```diff
@@ -35343,6 +35343,53 @@ static void test_wolfSSL_PEM_read(void)
...
+    AssertNotNull(ctx = EVP_CIPHER_CTX_new());
+    AssertIntEQ(EVP_EncryptInit_ex(ctx, EVP_aes_128_gcm(), NULL, NULL, NULL), 1);
+    AssertIntEQ(EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv), 1);
+    AssertIntEQ(EVP_EncryptUpdate(ctx, NULL, &len, aad, 1), 1);
+    AssertIntEQ(EVP_EncryptUpdate(ctx, NULL, &len, aad + 1, sizeof(aad) - 1), 1);
+    AssertIntEQ(EVP_EncryptUpdate(ctx, out2Part, &len, cleartext, sizeof(cleartext)), 1);
+    AssertIntEQ(EVP_EncryptFinal_ex(ctx, out2Part, &len), 1);
+    AssertIntEQ(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, outTag2Part), 1);
+    EVP_CIPHER_CTX_free(ctx);
+
```
Comment (SparkiDev) [inline tests/api.c:35385]: Add decrypt version of test as well.
Would like to see a test that reuses the AAD as well.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Review by SparkiDev — APPROVED** [review]

**Review by toddouska — APPROVED** [review]

**Pull request merged**